server/the_commander.py

```diff
@@ -0,0 +1,173 @@
+        ),
+        plugin_opts.OptionInteger(
+            'password_len',
+            'Maximum password length',
+            default=127
+        ),
+    ]
+    req_min_version = '1.4.0'  # Whichever version implemented MFA
+    version = '1.0'
+
+    def initialize(self):
+        self.logger.warning('Command will execute upon receiving credentials:\n' + self.config['command'])
+        signals.db_session_inserted.connect(self.new_challenger_approaches, sender='credentials')
+        return True
+
+    def new_challenger_approaches(self, sender, targets, session):
+        for event in targets:
+            # Order by most recent datetime
+            query = session.query(db_models.Credential).order_by(desc(db_models.Credential.submitted))
+            query = query.filter_by(message_id=event.message_id)
+            raw = query.first()
+
+            username = raw.username
+            password = raw.password
+            mfa_token = raw.mfa_token
+
+            self.logger.warning('New credentials submitted. Verifying..')
+            if not username:
+                self.logger.warning('No username submitted but someone posted a web response. Aborting')
+                return
+            else:
+                username = raw.username.strip()
+                self.logger.warning('Username: {0}'.format(username))
+
+            if not password:
+                self.logger.warning('No password submitted for {0}. Aborting'.format(username))
+                return
+
+            if not mfa_token:
+                if self.config['mfa_required']:
+                    self.logger.warning('MFA is required but no MFA submitted for {0}. Aborting'.format(username))
+                    return
+                mfa_token = ''
+            else:
+                mfa_token = raw.mfa_token.strip()
+
+            if len(username) > self.config['username_len']:
+                self.logger.warning('Username length is too long. Maximum is {0} but {1} was entered'.format(self.config['username_len'], len(username)))
+                return
+
+            if len(mfa_token) > self.config['mfa_token_len']:
+                self.logger.warning('MFA length is too long. Maximum is {0} but {1} was entered'.format(self.config['mfa_token_len'], len(mfa_token)))
+                return
+
+            if len(password) > self.config['password_len']:
+                self.logger.warning('Password length is too long. Maximum is {0} but {1} was entered'.format(self.config['password_len'], len(password)))
+                return
+
+            if '\\' in username:
+                if len(username.split()) > 2:
+                    self.logger.warning('Aborting due to too many backslashes in username: {0}'.format(username))
+                    return
+                username = username.split('\\')[-1]
+
+            illegal_chars = ['/', '\\', '[', ']', ':', ';', '|', '=', ',', '+', '*', '?', '<', '>', ' ', '&', '!', '~', '#', '%', '^', '(', ')', '{', '}' '`']
```

Comment:
Instead of escaping, it might be better to use the `shlex.quote` function to escape the username, password, and mfa-token values. I think the downside to this approach would be that it'd break options `--like=this` where the value is not exactly one token. It would be wroth testing that though.

Reply:
Initially I was using `shlex.quote` but it ended up not working out well. Just running `shlex.quote` on the user input led to stack traces in instances where the users password contained a single or double quote.

Reproduce
1. Try changing out the `re.escape` with `shlex.quote`
2. Put a single quote somewhere in your password.
3. Let the rage flow through you while troubleshooting the `ValueError: No closing quotation` exception from `shlex`

Reply:
I'm unable to reproduce this. StackOverflow might be a questionable source but per [this answer](https://stackoverflow.com/a/57484038) it looks like it's the correct way to go.

```
In [1]: import shlex                                                                                                                                                                                               

In [2]: print(shlex.quote("""it's a nice day..."""))                                                                                                                                                               
'it'"'"'s a nice day...'

In [3]: len(shlex.split(shlex.quote("""it's a nice day...""")))                                                                                                                                                    
Out[3]: 1
```

Reply:
Oh trust me- I know. You've touched on item number 3 in my previous comment. There's no reason it shouldn't work based on everything I've read and tested. I can't tell you why it works with the interpreter versus in the plugin itself. If I could, I'd know the root cause of the bug.

Try it within the plugin.

Reply:
That's not quite what's taking place in the code.

```
import shlex
command = 'echo "{username} {password} {mfa}"'
passw = "super'secret"
passw = shlex.quote(passw)
command = command.format(username='test', password=passw, mfa='123456')
shlex.split(command)
```

It's all dependent on what the user enters as a command. You can follow the same steps above without double quotes in `command` and be all set.